sympy/core/mod.py

```diff
@@ -107,6 +107,17 @@ def doit(p, q):
             elif (qinner*(q + qinner)).is_nonpositive:
                 # |qinner| < |q| and have different sign
                 return p
+        elif any(isinstance(arg, cls) for arg in p.args):
```

Comment:
Not all operations commute with mod. There needs to be a check on the type of `p` here.